README.md
perfmjs-redis-cluster
=======
perfmjs-redis-cluster is a Node.js version of Redis-rb-cluster. V1.1.1

Features:
=======


Test Result
=======
TODO, coming soon!


How to Use
-------
>npm install perfmjs-redis-cluster

License
-------

Copyright 2011 Tony

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

History.md
1.1.2 / 2015-08-06
 * improved: 提升perfmjs-node版本到1.4.2

1.1.1 / 2015-01-6
 * improved: 提升perfmjs-node版本到1.4.1

1.1.0 / 2014-09-19
 * improved: 提升perfmjs-node版本到1.3.5
 * improved: 提升redis版本到0.12.1
 * fixed: redis出错的时候不调用redisClient.end();

1.0.8 / 2014-09-10
==================
 * improved: 重构核心代码

1.0.7 / 2014-09-02
==================
 * improved: 增加pub/sub的测试用例

1.0.6 / 2014-08-31
==================
 * improved: 提升perfmjs-node版本到1.3.1

1.0.4 / 2014-08-22
==================
 * improved: 提升perfmjs-node版本到1.2.9

1.0.3 / 2014-07-29
==================
 * Add: add redis connection pool

1.0.2 / 2014-07-25
==================
 * Remove: remove unuseful log

1.0.1 / 2014-07-23
==================
 * the first stable version


0.0.2 / 2014-07-23
==================
 * fix: add return at init function

0.0.1 / 2014-07-23
==================
 * Init: init version


test/perfmjs-spec.js
/**
 * npm install jasmine-node -g
 * 在项目根路径下运行测试的方法： jasmine-node test/
 * 参考：https://www.npmjs.org/package/jasmine-node
 * Created by tony on 2014/7/24
 */
describe("测试perfmjs-redis-cluster", function () {
    beforeEach(function() {
        require("perfmjs-node");
    });
    it("应能测试通过redis#set方法", function() {
        require('../lib');
        perfmjs.ready(function($$, app) {
            app.registerAndStart($$.redisCluster);
            var redisCluster = $$.redisCluster.instance.initStartupOptions($$.sysConfig.config.get("redis.clusterNodes"));
            redisCluster.set('foo', 'test2', function(err, reply, redisClient) {
                if (err) {
                    $$.logger.error('error: ' + err.message);
                    return;
                }
                redisClient.get('foo', function (err, reply) {
                    if (err) {
                        $$.logger.info('redis error: ' + err);
                        return;
                    }
                    expect(1).toEqual(1);
                    $$.logger.info('reply:' + reply);
                    $$.logger.info("redisClient connected123:" + redisClient.connected);
                    redisClient.quit();
                    redisClient.end();
                    setTimeout(function() {
                        redisCluster.set('foo', 'test');
                        redisClient.send_command('asking',[], function(err, reply) {
                            if (err) {
                                $$.logger.error("error comming " + err.message);
                                return;
                            }
                            $$.logger.info('reply11:' + reply);
                        });
                        $$.logger.info("redisClient connected:" + redisClient.connected);
                    }, 5000);
                });
            });
        });
    });
    it("应能测试通过redis#pub/sub方法", function() {
        perfmjs.ready(function($$, app) {
            var redisHost = "192.168.66.47";
            var startNodes = [
                {host: redisHost, port: 7000},
                {host: redisHost, port: 7001},
                {host: redisHost, port: 7002}
            ];
            var channel = '/realtimeApp/dataChange/kc';
            app.stop($$.redisCluster.getName());
            app.unregister($$.redisCluster.getName());
            app.registerAndStart($$.redisCluster);
            var redisCluster = $$.redisCluster.instance.initStartupOptions(startNodes);
            console.log("keySlots:" + redisCluster._keyslot(channel));
            //$$.utils.nextTick(function() {
            setInterval(function() {
                redisCluster.publish(channel, '316');
            }, 2000);
            //});
            redisCluster.subscribe(channel, function(err, reply, redis) {
                if (err) {
                    $$.logger.error('error Occurred at redis subscribe: ' + err.message);
                    return;
                }
                redis.on("message", function (channel, message) {
                    console.log('channel:' + channel + "/message=" + message);
                });
            });
            expect(1).toEqual(1);
        });
    });
});

start.js
var perfmjs = require('perfmjs-node');
perfmjs.ready(function ($$, app) {
    console.log('hello, perfmjs-redis-cluster!');
    var channel = '/realtimeApp/dataChange/kc';
    app.registerAndStart(require('./lib'));
    var redisCluster = $$.redisCluster.instance.initStartupOptions($$.sysConfig.config.get("redis.clusterNodes"));
    console.log("keySlots:" + redisCluster._keyslot(channel));
    $$.utils.nextTick(function() {
        redisCluster.publish(channel, '316');
    });
    redisCluster.subscribe(channel, function (err, reply, redis) {
        if (err) {
            $$.logger.error('error Occurred at redis subscribe: ' + err.message);
            return;
        }
        redis.on("message", function (channel, message) {
            console.log('channel:' + channel + "/message=" + message);
        });
    });
});

lib/crc16.js
var crc16tab= [
    0x0000,0x1021,0x2042,0x3063,0x4084,0x50a5,0x60c6,0x70e7,
    0x8108,0x9129,0xa14a,0xb16b,0xc18c,0xd1ad,0xe1ce,0xf1ef,
    0x1231,0x0210,0x3273,0x2252,0x52b5,0x4294,0x72f7,0x62d6,
    0x9339,0x8318,0xb37b,0xa35a,0xd3bd,0xc39c,0xf3ff,0xe3de,
    0x2462,0x3443,0x0420,0x1401,0x64e6,0x74c7,0x44a4,0x5485,
    0xa56a,0xb54b,0x8528,0x9509,0xe5ee,0xf5cf,0xc5ac,0xd58d,
    0x3653,0x2672,0x1611,0x0630,0x76d7,0x66f6,0x5695,0x46b4,
    0xb75b,0xa77a,0x9719,0x8738,0xf7df,0xe7fe,0xd79d,0xc7bc,
    0x48c4,0x58e5,0x6886,0x78a7,0x0840,0x1861,0x2802,0x3823,
    0xc9cc,0xd9ed,0xe98e,0xf9af,0x8948,0x9969,0xa90a,0xb92b,
    0x5af5,0x4ad4,0x7ab7,0x6a96,0x1a71,0x0a50,0x3a33,0x2a12,
    0xdbfd,0xcbdc,0xfbbf,0xeb9e,0x9b79,0x8b58,0xbb3b,0xab1a,
    0x6ca6,0x7c87,0x4ce4,0x5cc5,0x2c22,0x3c03,0x0c60,0x1c41,
    0xedae,0xfd8f,0xcdec,0xddcd,0xad2a,0xbd0b,0x8d68,0x9d49,
    0x7e97,0x6eb6,0x5ed5,0x4ef4,0x3e13,0x2e32,0x1e51,0x0e70,
    0xff9f,0xefbe,0xdfdd,0xcffc,0xbf1b,0xaf3a,0x9f59,0x8f78,
    0x9188,0x81a9,0xb1ca,0xa1eb,0xd10c,0xc12d,0xf14e,0xe16f,
    0x1080,0x00a1,0x30c2,0x20e3,0x5004,0x4025,0x7046,0x6067,
    0x83b9,0x9398,0xa3fb,0xb3da,0xc33d,0xd31c,0xe37f,0xf35e,
    0x02b1,0x1290,0x22f3,0x32d2,0x4235,0x5214,0x6277,0x7256,
    0xb5ea,0xa5cb,0x95a8,0x8589,0xf56e,0xe54f,0xd52c,0xc50d,
    0x34e2,0x24c3,0x14a0,0x0481,0x7466,0x6447,0x5424,0x4405,
    0xa7db,0xb7fa,0x8799,0x97b8,0xe75f,0xf77e,0xc71d,0xd73c,
    0x26d3,0x36f2,0x0691,0x16b0,0x6657,0x7676,0x4615,0x5634,
    0xd94c,0xc96d,0xf90e,0xe92f,0x99c8,0x89e9,0xb98a,0xa9ab,
    0x5844,0x4865,0x7806,0x6827,0x18c0,0x08e1,0x3882,0x28a3,
    0xcb7d,0xdb5c,0xeb3f,0xfb1e,0x8bf9,0x9bd8,0xabbb,0xbb9a,
    0x4a75,0x5a54,0x6a37,0x7a16,0x0af1,0x1ad0,0x2ab3,0x3a92,
    0xfd2e,0xed0f,0xdd6c,0xcd4d,0xbdaa,0xad8b,0x9de8,0x8dc9,
    0x7c26,0x6c07,0x5c64,0x4c45,0x3ca2,0x2c83,0x1ce0,0x0cc1,
    0xef1f,0xff3e,0xcf5d,0xdf7c,0xaf9b,0xbfba,0x8fd9,0x9ff8,
    0x6e17,0x7e36,0x4e55,0x5e74,0x2e93,0x3eb2,0x0ed1,0x1ef0
];

exports = module.exports = function (str) {
    var i = 0,
        len = str.length,
        crc = 0;
    for (i = 0; i < len; i++)
        crc = (crc<<8) ^ crc16tab[((crc>>8) ^ str.charCodeAt(i))&0x00FF];
    return crc;
}


lib/commands.js
// This file was generated by ./generate_commands.js on Wed Apr 23 2014 14:51:21 GMT-0700 (PDT)
module.exports = [
    "append",
    "auth",
    "bgrewriteaof",
    "bgsave",
    "bitcount",
    "bitop",
    "bitpos",
    "blpop",
    "brpop",
    "brpoplpush",
    "client kill",
    "client list",
    "client getname",
    "client pause",
    "client setname",
    "config get",
    "config rewrite",
    "config set",
    "config resetstat",
    "dbsize",
    "debug object",
    "debug segfault",
    "decr",
    "decrby",
    "del",
    "discard",
    "dump",
    "echo",
    "eval",
    "evalsha",
    "exec",
    "exists",
    "expire",
    "expireat",
    "flushall",
    "flushdb",
    "get",
    "getbit",
    "getrange",
    "getset",
    "hdel",
    "hexists",
    "hget",
    "hgetall",
    "hincrby",
    "hincrbyfloat",
    "hkeys",
    "hlen",
    "hmget",
    "hmset",
    "hset",
    "hsetnx",
    "hvals",
    "incr",
    "incrby",
    "incrbyfloat",
    "info",
    "keys",
    "lastsave",
    "lindex",
    "linsert",
    "llen",
    "lpop",
    "lpush",
    "lpushx",
    "lrange",
    "lrem",
    "lset",
    "ltrim",
    "mget",
    "migrate",
    "monitor",
    "move",
    "mset",
    "msetnx",
    "multi",
    "object",
    "persist",
    "pexpire",
    "pexpireat",
    "pfadd",
    "pfcount",
    "pfmerge",
    "ping",
    "psetex",
    "psubscribe",
    "pubsub",
    "pttl",
    "publish",
    "punsubscribe",
    "quit",
    "end",
    "randomkey",
    "rename",
    "renamenx",
    "restore",
    "rpop",
    "rpoplpush",
    "rpush",
    "rpushx",
    "sadd",
    "save",
    "scard",
    "script exists",
    "script flush",
    "script kill",
    "script load",
    "sdiff",
    "sdiffstore",
    "select",
    "set",
    "setbit",
    "setex",
    "setnx",
    "setrange",
    "shutdown",
    "sinter",
    "sinterstore",
    "sismember",
    "slaveof",
    "slowlog",
    "smembers",
    "smove",
    "sort",
    "spop",
    "srandmember",
    "srem",
    "strlen",
    "subscribe",
    "sunion",
    "sunionstore",
    "sync",
    "time",
    "ttl",
    "type",
    "unsubscribe",
    "unwatch",
    "watch",
    "zadd",
    "zcard",
    "zcount",
    "zincrby",
    "zinterstore",
    "zlexcount",
    "zrange",
    "zrangebylex",
    "zrangebyscore",
    "zrank",
    "zrem",
    "zremrangebylex",
    "zremrangebyrank",
    "zremrangebyscore",
    "zrevrange",
    "zrevrangebyscore",
    "zrevrank",
    "zscore",
    "zunionstore",
    "scan",
    "sscan",
    "hscan",
    "zscan"
];


lib/redis-cluster.js
/**
 * This is a Node.js version of Redis-rb-cluster.
 * Created by tony on 2014/7/22.
 */
require("perfmjs-node");
perfmjs.plugin('redisCluster', function($$) {
    $$.base("redisCluster", {
        init: function(eventProxy) {
            this.option('eventProxy', eventProxy).on($$.sysConfig.events.moduleIsReady, function() {$$.logger.info("RedisCluster is Ready!");});
            return this;
        },

        /**
         * 初始化参数
         * @param startupNodes
         * @param options
         */
        initStartupOptions: function(startupNodes, options) {
            this.option('startupNodes', startupNodes);
            options = options || {};
            if (!options['connTimeout']) {
                options['connTimeout'] = this.option('redisClusterDefaultTimeout');
            }
            this.option('connOption', options);
            this._addCommandsFunc();
            return this;
        },

        /**
         * 获取redis client
         * @param key e.g. key is 'foo' when command: set('foo', bar')
         * @param callback
         * @param usingStartupNodeIndex e.g. 0
         */
        _getRedisClient: function(key, callback, usingStartupNodeIndex) {
            var self = this, connFromPool = this._getConnectionFromPool(key);
            if (connFromPool) {
                callback(connFromPool);
                return;
            }
            usingStartupNodeIndex = usingStartupNodeIndex || 0;
            //只获取1次slots
            for (var i = usingStartupNodeIndex; i < this.option('startupNodes').length; i++) {
                var node = this.option('startupNodes')[i];
                this._getHashSlots(node['host'], node['port'], function(err) {
                    if (err) {
                        self._getRedisClient(key, callback, ++usingStartupNodeIndex);
                        return;
                    }
                    //get redis client
                    var slotNode = self._getNodeBySlot(self._keyslot(key));
                    if (!slotNode) {
                        self._getRedisClient(key, callback, ++usingStartupNodeIndex);
                        return;
                    }
                    //$$.logger.info('key:' + key + ' matched slotNode: ' + slotNode['name']);
                    callback(self._setConnectionPool(key, self._getRedisLink(slotNode['host'], slotNode['port'])));
                });
                break;
            }
        },

        /**
         * set connection pool
         * @param key
         * @param redisClient
         * @private
         */
        _setConnectionPool: function(key, redisClient) {
            if (!key) {
                return redisClient;
            }
            if ($$.utils.keys(this.option('connectionPool')).length > this.option('maxConnections')) {
                //do nothing
                return redisClient;
            }
            return this.option('connectionPool')[key] = redisClient;
        },

        /**
         * get connection from connection pool
         * @param key
         * @private
         */
        _getConnectionFromPool: function(key) {
            if (key && this.option('connectionPool')[key]) {
                if (this.option('connectionPool')[key]['connected']) {
                    return this.option('connectionPool')[key];
                } else {
                    delete this.option('connectionPool')[key];
                }
            }
        },

        /**
         * Return the hash slot from the key
         */
        _keyslot: function(key) {
            //Only hash what is inside {...} if there is such a pattern in the key.
            //Note that the specification requires the content that is between
            //the first { and the first } after the first {. If we found {} without
            //nothing in the middle, the whole key is hashed as usually.
            if (!key) {
                return 0;
            }
            var s = key.indexOf('{');
            if(s > 0){
                var e = key.indexOf('}',s+1);
                if (e && e != s+1){
                    key = key.substring(s,e);
                }
            }
            return this.option('crc16')(key)&(this.option('redisClusterHashSlots') - 1);
        },

        _getNodeBySlot: function(slot) {
            var node;
            var specialNode = $$.joquery.newInstance(this.option('slots')).where(function(slotNode, index) {
                if ($$.utils.toNumber(slotNode['startSlot']) <= $$.utils.toNumber(slot) && $$.utils.toNumber(slotNode['endSlot']) >= $$.utils.toNumber(slot)) {
                    return true;
                }
            },true).toArray();
            if(specialNode[0]) {
                node = specialNode[0]['master'];
            } else {
                $$.logger.info("slot not exist: " + slot);
            }
            return node;
        },

        _getRedisLink: function(host, port){
            //FIXME save connect to this.option('connects')
            var redisClient = this.option('redis').createClient(port, host, this.option('connOptions'));
            redisClient.on('error',function(err){
                console.error("error on client " + err);
            });
            return redisClient;
        },

        _getHashSlots: function(host, port, callback) {
            var self = this, redisLink = this._getRedisLink(host, port);
            if (!redisLink) {
                callback($$.utils.error('get Redis Client fail on host:' + host + ", port:" + port));
                return false;
            }
            redisLink.send_command('cluster', ['nodes'], function(err, reply) {
                if (err || (typeof reply == 'undefined')) {
                    callback(err);
                    return;
                }
                //console.log("All nodes:" + reply);
                var connectedSlots = $$.joquery.newInstance(self._parseNodesToSlots(reply)).select(function(item) {
                    return item['master']['connected'] || item['slave']['connected'];
                }).orderBy(function(item) {
                    return item['startSlot'];
                }).toArray();
                self.option('slots', connectedSlots);
                var connectedSlotNodes = [];
                $$.utils.forEach(connectedSlots, function(item, index) {
                    if (item['master']['connected']) {
                        connectedSlotNodes[connectedSlotNodes.length] = item['master'];
                    }
                    if (item['slave']['connected']) {
                        connectedSlotNodes[connectedSlotNodes.length] = item['slave'];
                    }
                });
                self.option('nodes', connectedSlotNodes);
                self._populateStartupNodes();
                self.option('refresh_table_asap', false);
                //回调函数
                callback();
            });
            return true;
        },

        _parseNodesToSlots: function(nodeReply) {
            var allSlots = [], address, data, l, lines, nodes, result, _i, _len, _ref, nodes = [], result = [], lines = nodeReply.split('\n');
            for (_i = 0, _len = lines.length; _i < _len; _i++) {
                var nodeInfo = {id:'', host:'', port:'', name:'', isMaster:false, refMater:'', connected:false, slot:''};
                l = lines[_i];
                data = l.split(" ");
                if (!data || data.length < 8) {
                    continue;
                }
                address = data[1].split(':');
                nodeInfo.id = data[0];
                nodeInfo.host = address[0];
                nodeInfo.port = address[1];
                nodeInfo.name = data[1];
                nodeInfo.isMaster = (data[2].indexOf('master')>=0)?true:false;
                nodeInfo.refMater = (data[3]=='-')?'':data[3];
                if (data[7] === 'connected') {
                    nodeInfo.connected = true;
                }
                if (data[8] && (data[8].indexOf('-') > 0)) {
                    nodeInfo['slot'] = data[8];
                }
                result[result.length] = nodeInfo;
            }
            var masterNodes = $$.joquery.newInstance(result).select(function(node) {
                return node['isMaster'];
            }).toArray();
            var slaveNodes = $$.joquery.newInstance(result).select(function(node) {
                return !node['isMaster'] && (node['refMater'].length > 0);
            }).toArray();
            var slotNodes = $$.joquery.newInstance(result).select(function(node) {
                return node['isMaster'] && node['slot'].indexOf('-') > 0;
            }).toArray();
            $$.utils.forEach(slotNodes, function(node, index) {
                var slots = node['slot'].split('-');
                var thisSlot = {startSlot: slots[0], endSlot: slots[1], master:{host:node['host'], port:node['port'], name:node['name'], connected:node['connected']}, slave:{host:'', port:0, name:'', connected:false}};
                var slaveNode = $$.joquery.newInstance(slaveNodes).where(function(item) {
                    return (node['id'] === item['refMater']);
                }).toArray();
                if (slaveNode.length > 0) {
                    thisSlot['slave']['host'] = slaveNode[0]['host'];
                    thisSlot['slave']['port'] = slaveNode[0]['port'];
                    thisSlot['slave']['name'] = slaveNode[0]['name'];
                    thisSlot['slave']['connected'] = slaveNode[0]['connected'];
                }
                allSlots[allSlots.length] = thisSlot;
            });
            return allSlots;
        },

        /**
         * Flush the cache, mostly useful for debugging when we want to force
         * redirection.
         * @private
         */
        _flushSlotsCache: function() {
            this.option('nodes', []);
            this.option('slots', []);
        },

        _populateStartupNodes: function() {
            var self = this;
            $$.utils.forEach(this.option('startupNodes'), function(node, index) {
                node['name'] = self._getNodeName(node.host, node.port);
            });
            $$.utils.forEach(this.option('nodes'), function(node, index) {
                self.option('startupNodes')[self.option('startupNodes').length] = node;
            });
            var uniqNodes = $$.joquery.newInstance(this.option('startupNodes')).distinct(function(node) {
                return node['name'];
            }).toArray();
            this.option('startupNodes', uniqNodes);
        },

        /**
         * Return the first key in the command arguments.
         Currently we just return argv[1], that is, the first argument
         after the command name.
         This is indeed the key for most commands, and when it is not true
         the cluster redirection will point us to the right node anyway.
         For commands we want to explicitly bad as they don't make sense
         in the context of cluster, nil is returned.
         * @param args
         * @returns {*}
         * @private
         */
        _getKeyFromCommand: function(args){
            var command = args[0].toLowerCase();
            if(command === 'info' || command === 'multi' || command === 'exec' || command === 'slaveof' || command === 'config' || command === 'shutdown'){
                return undefined;
            } else if (command === 'eval'|| command === 'evalsha'){
                return args[3] + "";
            } else {
                return args[1] + "";
            }
        },

        _getNodeName: function(host, port) {
            return host + ":" + port;
        },

        _addCommandsFunc: function() {
            var self = this;
            $$.utils.forEach(this.option('commands'), function(command, index) {
                $$[self.name].prototype[command.toUpperCase()]  = self._addDynamicFunc(command, function() {
                    var args = Array.prototype.slice.call(arguments) || [];
                    args.unshift(command);
                    var params = args.slice(1) || [];
                    var commandKey = self._getKeyFromCommand(args);
                    var callback = args[args.length - 1];
                    if (callback && typeof callback === "function") {
                        params = params.slice(0, params.length - 1) || [];
                    } else {
                        callback = undefined;
                    }
                    self._getRedisClient(commandKey, function(redisClient) {
                        if (!redisClient) {
                            if (callback) {
                                callback(err);
                            }
                            return;
                        }
                        redisClient[command](params, function(err, reply) {
                            if (err) {
                                $$.logger.error('error on command:' + command + ', error msg:' + err.message);
                            }
                            if (callback) {
                                callback(err, reply, redisClient);
                            }
                        });
                    });
                });
            });
        },

        /**
         * 动态增加函数
         * @param funcName
         * @param fn
         * @private
         */
        _addDynamicFunc: function(funcName, fn) {
            if ($$[this.name].prototype[funcName]) {
                return $$[this.name].prototype[funcName];
            }
            return $$[this.name].prototype[funcName] = fn;
        },
        end: 0
    });
    $$.redisCluster.defaults = {
        redis: require('redis'),
        crc16: require('./crc16'),
        commands: require('./commands'),
        startupNodes: [], //[{host:'*.*.*.*',port:7000,name:'*.*.*.*:7000'}]
        connOption: {}, //其它连接参数，如connTimeout
        maxConnections: 30,
        redisClusterHashSlots: 16384,
        redisClusterRequestTTL: 5,  //redirect times when error or moved
        redisClusterDefaultTimeout: 5,
        slots: [], //slot num(0~16383) ---> [{startSlot:0, endSlot:5460, master:{host:'*.*.*.*', ip:7000, name:'*.*.*.*:7000'}}, slave:{host:'*.*.*.*', ip:7000, name:'*.*.*.*:7003'}}]
        nodes: [], //id --->[{host,port,name}]
        connectionPool: {}, //连接池，e.g. {'key': 'redis连接实例'}
        eventProxy: {},
        end: 0
    };
    /*for Node.js begin*/
    if (typeof module !== 'undefined' && module.exports) {
        exports = module.exports = perfmjs.redisCluster;
    }
    /*for Node.js end*/
});